src/provider/provider_tracking.c

```diff
@@ -629,47 +652,85 @@ static umf_result_t trackingOpenIpcHandle(void *provider, void *providerIpcData,
     umf_result_t ret = UMF_RESULT_SUCCESS;
 
     assert(p->hUpstream);
+    assert(p->hIpcMappedCache);
+
+    umf_ipc_data_t *ipcUmfData = getIpcDataFromIpcHandle(providerIpcData);
+
+    // Compiler may add paddings to the ipc_mapped_handle_cache_key_t structure
+    // so we need to zero it out to avoid false cache miss.
+    ipc_mapped_handle_cache_key_t key = {0};
+    key.remote_base_ptr = ipcUmfData->base;
+    key.local_provider = provider;
+    key.remote_pid = ipcUmfData->pid;
 
-    ret = umfMemoryProviderOpenIPCHandle(p->hUpstream, providerIpcData, ptr);
+    ipc_mapped_handle_cache_value_t *cache_entry = NULL;
+    ret = umfIpcHandleMappedCacheGet(p->hIpcMappedCache, &key,
+                                     ipcUmfData->handle_id, &cache_entry);
     if (ret != UMF_RESULT_SUCCESS) {
-        LOG_ERR("upstream provider failed to open IPC handle");
+        LOG_ERR("failed to get cache entry");
         return ret;
     }
-    size_t bufferSize = getDataSizeFromIpcHandle(providerIpcData);
-    ret = umfMemoryTrackerAdd(p->hTracker, p->pool, *ptr, bufferSize);
-    if (ret != UMF_RESULT_SUCCESS) {
-        LOG_ERR("failed to add IPC region to the tracker, ptr=%p, size=%zu, "
-                "ret = %d",
-                *ptr, bufferSize, ret);
-        if (umfMemoryProviderCloseIPCHandle(p->hUpstream, *ptr, bufferSize)) {
-            LOG_ERR("upstream provider failed to close IPC handle, ptr=%p, "
-                    "size=%zu",
-                    *ptr, bufferSize);
+
+    assert(cache_entry != NULL);
+
+    void *mapped_ptr = NULL;
+    utils_atomic_load_acquire(&(cache_entry->mapped_base_ptr), &mapped_ptr);
+    if (mapped_ptr == NULL) {
+        utils_mutex_lock(&(cache_entry->mmap_lock));
```

Comment:
Please create separate function, to open and update cache entry. This will improve readability of the code (you can have better errorhandle/return flow in separate function)

Reply:
Done